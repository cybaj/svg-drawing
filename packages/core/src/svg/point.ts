import type { PointObject } from '../types'

/**
 * Maximum number of decimal places is 5
 *
 * @todo Changeable parameter
 */
const coeffecient = 100000
const roundNumber = (num: number) => Math.round(num * coeffecient) / coeffecient

export class Point implements PointObject {
  private _x: number
  private _y: number
  constructor(x: number, y: number) {
    this._x = roundNumber(x)
    this._y = roundNumber(y)
  }

  public get x(): number {
    return this._x
  }

  public get y(): number {
    return this._y
  }

  public get degrees(): number {
    return Math.atan2(this._y, this._x) * (180 / Math.PI)
  }

  public get absoluteValue(): number {
    return Math.sqrt(Math.pow(this._x, 2.0) + Math.pow(this._y, 2.0))
  }

  public scale(r: number): Point {
    return new Point(this._x * r, this._y * r)
  }

  public scaleX(r: number): Point {
    return new Point(this._x * r, this._y)
  }

  public scaleY(r: number): Point {
    return new Point(this._x, this._y * r)
  }

  public add(p: PointObject): Point {
    return new Point(this._x + p.x, this._y + p.y)
  }

  public sub(p: PointObject): Point {
    return new Point(this._x - p.x, this._y - p.y)
  }

  public translate(p: PointObject): void {
    this._x += p.x
    this._y += p.y
  }

  public eql(p: PointObject): boolean {
    return this._x === p.x && this._y === p.y
  }

  public clone(): Point {
    return new Point(this._x, this._y)
  }

  public toJson(): PointObject {
    return {
      x: this._x,
      y: this._y,
    }
  }

  public static fromVector({
    value,
    degrees,
  }: {
    value: number
    degrees: number
  }) {
    const theta = (degrees * Math.PI) / 180
    return new Point(Math.cos(theta) * value, Math.sin(theta) * value)
  }
}
