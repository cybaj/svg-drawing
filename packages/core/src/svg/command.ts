import { CommandType, PointObject } from '../types'
import { roundUp } from '../utils'
import { Point } from './point'

export const COMMAND_TYPE: { [name: string]: CommandType } = {
  MOVE: 'M', // M 0 0
  MOVE_RELATIVE: 'm', // m 0 0
  LINE: 'L', // L 1 1
  LINE_RELATIVE: 'l', // l 1 1
  CUBIC_BEZIER_CURVE: 'C', // C 1 1 2 2 3 3
  CUBIC_BEZIER_CURVE_RELATIVE: 'c', // c 1 1 2 2 3 3
  CLOSE: 'Z', // Z, z
  HORIZONTAL: 'H', // H 10
  HORIZONTAL_RELATIVE: 'h', // h 10
  VERTICAL: 'V', // V 20
  VERTICAL_RELATIVE: 'v', // v 20
  ARC_CURVE: 'A', // A 6 4 10 0 1 14 10
  ARC_CURVE_RELATIVE: 'a', // A 6 4 10 0 1 14 10
  QUADRATIC_CURVE: 'Q', // Q 10 60 10 30
  QUADRATIC_CURVE_RELATIVE: 'q', // q 10 60 10 30
  SHORTCUT_CURVE: 'S', // S 10 60 10 30
  SHORTCUT_CURVE_RELATIVE: 's', // s 10 60 10 30
} as const

// TODO: compatible COMMAND_TYPE
export class Command {
  public type: CommandType
  public value: number[]
  // TODO: Convert data format to number array.
  constructor(type: CommandType, value: number[] = []) {
    this.value = value
    this.type = type
  }

  public set cr(po: Point | undefined) {
    if (!po) return
    if (!this.isCubicBezierCurve) {
      return
    }
    this.value.splice(2, 1, po.x)
    this.value.splice(3, 1, po.y)
  }
  public get cr(): Point | undefined {
    if (!this.isCubicBezierCurve) {
      return undefined
    }
    const [x, y] = this.value.slice(2, 4)
    return new Point(x, y)
  }
  public set cl(po: Point | undefined) {
    if (!po) return
    if (!this.isCurve) {
      return
    }
    this.value.splice(0, 1, po.x)
    this.value.splice(1, 1, po.y)
  }
  public get cl(): Point | undefined {
    if (!this.isCurve) {
      return undefined
    }
    const [x, y] = this.value.slice(0, 2)
    return new Point(x, y)
  }

  public set point(po: Point | undefined) {
    if (!po) return
    this.value.splice(this.value.length - 2, 1, po.x)
    this.value.splice(this.value.length - 1, 1, po.y)
  }
  public get point(): Point | undefined {
    const xy = this.value.slice(this.value.length - 2)
    return xy.length === 2 ? new Point(xy[0], xy[1]) : undefined
  }

  public toString(): string {
    if (this.type === COMMAND_TYPE.CLOSE) return COMMAND_TYPE.CLOSE
    return `${this.type} ${this.value.map((v) => roundUp(v)).join(' ')}`
  }

  public scale(r: number): Command {
    const upd = new Command(
      this.type,
      this.value.map((p) => p * r)
    )
    return upd
  }

  public scaleX(r: number): Command {
    const point = this.point?.scaleX(r)
    const cl = this.cl?.scaleX(r)
    const cr = this.cr?.scaleX(r)
    return new Command(
      this.type,
      [cl, cr, point].reduce(
        (res: number[], po: Point | undefined) =>
          po ? [...res, po.x, po.y] : res,
        []
      )
    )
  }

  public scaleY(r: number): Command {
    const point = this.point?.scaleY(r)
    const cl = this.cl?.scaleY(r)
    const cr = this.cr?.scaleY(r)
    return new Command(
      this.type,
      [cl, cr, point].reduce(
        (res: number[], po: Point | undefined) =>
          po ? [...res, po.x, po.y] : res,
        []
      )
    )
  }

  public clone(): Command {
    return new Command(this.type, this.value.slice())
  }

  public get isCubicBezierCurve(): boolean {
    switch (this.type) {
      case COMMAND_TYPE.CUBIC_BEZIER_CURVE:
      case COMMAND_TYPE.CUBIC_BEZIER_CURVE_RELATIVE:
        return this.value.length === 6
      default:
        return false
    }
  }
  public get isCurve(): boolean {
    switch (this.type) {
      case COMMAND_TYPE.CUBIC_BEZIER_CURVE:
      case COMMAND_TYPE.CUBIC_BEZIER_CURVE_RELATIVE:
        return this.value.length === 6
      case COMMAND_TYPE.QUADRATIC_CURVE:
      case COMMAND_TYPE.QUADRATIC_CURVE_RELATIVE:
      case COMMAND_TYPE.SHORTCUT_CURVE:
      case COMMAND_TYPE.SHORTCUT_CURVE_RELATIVE:
        return this.value.length === 4
      default:
        return false
    }
  }

  public get isAbsolute(): boolean {
    return [
      COMMAND_TYPE.MOVE,
      COMMAND_TYPE.LINE,
      COMMAND_TYPE.CUBIC_BEZIER_CURVE,
      COMMAND_TYPE.ARC_CURVE,
      COMMAND_TYPE.QUADRATIC_CURVE,
      COMMAND_TYPE.SHORTCUT_CURVE_RELATIVE,
    ].includes(this.type)
  }
  public translate(p: PointObject) {
    if (!this.isAbsolute) return
    const po = new Point(p.x, p.y)
    this.point = this.point?.add(po)
    this.cr = this.cr?.add(po)
    this.cl = this.cl?.add(po)
  }
}
