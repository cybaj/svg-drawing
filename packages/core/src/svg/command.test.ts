import { Curve, Line, Move } from './command'
import { Point } from './point'

describe('Move', () => {
  it('new Move()', () => {
    expect(new Move(new Point(0, 0)).toString()).toBe('M0 0')
  })
  it.todo('point')
  it.todo('points')
  it.todo('values')
  it('clone', () => {
    const cmd = new Move(new Point(1, 1))
    const clone = cmd.clone()
    clone.points[0] = new Point(2, 2)
    expect(cmd.toString()).toBe('M1 1')
    expect(clone.toString()).toBe('M2 2')
  })
  it('scale', () => {
    expect(new Move(new Point(1, 1)).scale(2).toString()).toBe('M2 2')
  })
  it('scaleX', () => {
    expect(new Move(new Point(1, 1)).scaleX(2).toString()).toBe('M2 1')
  })
  it('scaleY', () => {
    expect(new Move(new Point(1, 1)).scaleY(2).toString()).toBe('M1 2')
  })
  it.todo('translate')
  it.todo('toRelative')
})

describe('Line', () => {
  it('new Line()', () => {
    expect(new Line(new Point(0, 0)).toString()).toBe('L0 0')
  })
  it.todo('point')
  it.todo('points')
  it.todo('values')
  it('clone', () => {
    const cmd = new Line(new Point(1, 1))
    const clone = cmd.clone()
    clone.points[0] = new Point(2, 2)
    expect(cmd.toString()).toBe('L1 1')
    expect(clone.toString()).toBe('L2 2')
  })
  it('scale', () => {
    expect(new Line(new Point(1, 1)).scale(2).toString()).toBe('L2 2')
  })
  it('scaleX', () => {
    expect(new Line(new Point(1, 1)).scaleX(2).toString()).toBe('L2 1')
  })
  it('scaleY', () => {
    expect(new Line(new Point(1, 1)).scaleY(2).toString()).toBe('L1 2')
  })
  it.todo('translate')
  it.todo('toRelative')
})

describe('Curve', () => {
  it('new Curve()', () => {
    expect(
      new Curve([
        new Point(0.25, 0.25),
        new Point(0.75, 0.25),
        new Point(1, 1),
      ]).toString()
    ).toBe('C0.25 0.25 0.75 0.25 1 1')
  })
  it.todo('point')
  it.todo('points')
  it.todo('values')
  it('clone', () => {
    const cmd = new Curve([
      new Point(0.25, 0.25),
      new Point(0.75, 0.25),
      new Point(1, 1),
    ])
    const clone = cmd.clone()
    clone.points[2] = new Point(1.5, 1)
    expect(cmd.point.x).toBe(1)
    expect(clone.point.x).toBe(1.5)
  })
  it('scale', () => {
    expect(
      new Curve([new Point(0.25, 0.25), new Point(0.75, 0.25), new Point(1, 1)])
        .scale(2)
        .toString()
    ).toBe('C0.5 0.5 1.5 0.5 2 2')
  })
  it('scaleX', () => {
    expect(
      new Curve([new Point(0.25, 0.25), new Point(0.75, 0.25), new Point(1, 1)])
        .scaleX(2)
        .toString()
    ).toBe('C0.5 0.25 1.5 0.25 2 1')
  })
  it('scaleY', () => {
    expect(
      new Curve([new Point(0.25, 0.25), new Point(0.75, 0.25), new Point(1, 1)])
        .scaleY(2)
        .toString()
    ).toBe('C0.25 0.5 0.75 0.5 1 2')
  })
  it.todo('translate')
  it.todo('toRelative')
})

describe('ArcCurve', () => {
  it.todo('new ArcCurve()')
  it.todo('point')
  it.todo('points')
  it.todo('values')
  it.todo('clone')
  it.todo('scale')
  it.todo('scaleX')
  it.todo('scaleY')
  it.todo('translate')
  it.todo('toRelative')
})

describe('Horizonal', () => {
  it.todo('new Horizonal()')
  it.todo('point')
  it.todo('points')
  it.todo('values')
  it.todo('clone')
  it.todo('scale')
  it.todo('scaleX')
  it.todo('scaleY')
  it.todo('translate')
  it.todo('toRelative')
})

describe('Vertical', () => {
  it.todo('new Vertical()')
  it.todo('point')
  it.todo('points')
  it.todo('values')
  it.todo('clone')
  it.todo('scale')
  it.todo('scaleX')
  it.todo('scaleY')
  it.todo('translate')
  it.todo('toRelative')
})

describe('RelativeMove', () => {
  it.todo('new RelativeMove()')
  it.todo('point')
  it.todo('points')
  it.todo('values')
  it.todo('clone')
  it.todo('scale')
  it.todo('scaleX')
  it.todo('scaleY')
  it.todo('translate')
  it.todo('toAbsolute')
})

describe('RelativeLine', () => {
  it.todo('new RelativeLine()')
  it.todo('point')
  it.todo('points')
  it.todo('values')
  it.todo('clone')
  it.todo('scale')
  it.todo('scaleX')
  it.todo('scaleY')
  it.todo('translate')
  it.todo('toAbsolute')
})

describe('RelativeCurve', () => {
  it.todo('new RelativeCurve()')
  it.todo('point')
  it.todo('points')
  it.todo('values')
  it.todo('clone')
  it.todo('scale')
  it.todo('scaleX')
  it.todo('scaleY')
  it.todo('translate')
  it.todo('toAbsolute')
})

describe('RelativeShortcutCurve', () => {
  it.todo('new RelativeShortcutCurve()')
  it.todo('point')
  it.todo('points')
  it.todo('values')
  it.todo('clone')
  it.todo('scale')
  it.todo('scaleX')
  it.todo('scaleY')
  it.todo('translate')
  it.todo('toAbsolute')
})

describe('RelativeQuadraticCurve', () => {
  it.todo('new RelativeQuadraticCurve()')
  it.todo('point')
  it.todo('points')
  it.todo('values')
  it.todo('clone')
  it.todo('scale')
  it.todo('scaleX')
  it.todo('scaleY')
  it.todo('translate')
  it.todo('toAbsolute')
})

describe('RelativeArcCurve', () => {
  it.todo('new RelativeArcCurve()')
  it.todo('point')
  it.todo('points')
  it.todo('values')
  it.todo('clone')
  it.todo('scale')
  it.todo('scaleX')
  it.todo('scaleY')
  it.todo('translate')
  it.todo('toAbsolute')
})

describe('RelativeHorizonal', () => {
  it.todo('new RelativeHorizonal()')
  it.todo('point')
  it.todo('points')
  it.todo('values')
  it.todo('clone')
  it.todo('scale')
  it.todo('scaleX')
  it.todo('scaleY')
  it.todo('translate')
  it.todo('toAbsolute')
})

describe('RelativeVertical', () => {
  it.todo('new RelativeVertical()')
  it.todo('point')
  it.todo('points')
  it.todo('values')
  it.todo('clone')
  it.todo('scale')
  it.todo('scaleX')
  it.todo('scaleY')
  it.todo('translate')
  it.todo('toAbsolute')
})

describe('Close', () => {
  it.todo('new Close()')
  it.todo('point')
  it.todo('points')
  it.todo('values')
  it.todo('clone')
  it.todo('scale')
  it.todo('scaleX')
  it.todo('scaleY')
  it.todo('translate')
})
