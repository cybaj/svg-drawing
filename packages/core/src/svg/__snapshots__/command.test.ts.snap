// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Curve new Curve() 1`] = `
Curve {
  "points": Array [
    Point {
      "_x": 0.25,
      "_y": 0.25,
    },
    Point {
      "_x": 0.75,
      "_y": 0.25,
    },
    Point {
      "_x": 1,
      "_y": 1,
    },
  ],
  "type": "C",
}
`;

exports[`Curve point 1`] = `
Point {
  "_x": 1,
  "_y": 1,
}
`;

exports[`Curve points 1`] = `
Array [
  Point {
    "_x": 0.25,
    "_y": 0.25,
  },
  Point {
    "_x": 0.75,
    "_y": 0.25,
  },
  Point {
    "_x": 1,
    "_y": 1,
  },
]
`;

exports[`Curve values 1`] = `
Array [
  0.25,
  0.25,
  0.75,
  0.25,
  1,
  1,
]
`;

exports[`Line new Line() 1`] = `
Line {
  "points": Array [
    Point {
      "_x": 0,
      "_y": 0,
    },
  ],
  "type": "L",
}
`;

exports[`Line point 1`] = `
Point {
  "_x": 0,
  "_y": 0,
}
`;

exports[`Line points 1`] = `
Array [
  Point {
    "_x": 0,
    "_y": 0,
  },
]
`;

exports[`Line values 1`] = `
Array [
  0,
  0,
]
`;

exports[`Move new Move() 1`] = `
Move {
  "points": Array [
    Point {
      "_x": 0,
      "_y": 0,
    },
  ],
  "type": "M",
}
`;

exports[`Move point 1`] = `
Point {
  "_x": 0,
  "_y": 0,
}
`;

exports[`Move points 1`] = `
Array [
  Point {
    "_x": 0,
    "_y": 0,
  },
]
`;

exports[`Move values 1`] = `
Array [
  Point {
    "_x": 0,
    "_y": 0,
  },
]
`;

exports[`RelativeCurve new RelativeCurve() 1`] = `
RelativeCurve {
  "points": Array [
    Point {
      "_x": 0.25,
      "_y": 0.25,
    },
    Point {
      "_x": 0.75,
      "_y": 0.25,
    },
    Point {
      "_x": 1,
      "_y": 1,
    },
  ],
  "type": "c",
}
`;

exports[`RelativeCurve point 1`] = `
Point {
  "_x": 1,
  "_y": 1,
}
`;

exports[`RelativeCurve points 1`] = `
Array [
  Point {
    "_x": 0.25,
    "_y": 0.25,
  },
  Point {
    "_x": 0.75,
    "_y": 0.25,
  },
  Point {
    "_x": 1,
    "_y": 1,
  },
]
`;

exports[`RelativeCurve values 1`] = `
Array [
  0.25,
  0.25,
  0.75,
  0.25,
  1,
  1,
]
`;

exports[`RelativeLine new RelativeLine() 1`] = `
RelativeLine {
  "points": Array [
    Point {
      "_x": 0,
      "_y": 0,
    },
  ],
  "type": "l",
}
`;

exports[`RelativeLine point 1`] = `
Point {
  "_x": 0,
  "_y": 0,
}
`;

exports[`RelativeLine points 1`] = `
Array [
  Point {
    "_x": 0,
    "_y": 0,
  },
]
`;

exports[`RelativeLine values 1`] = `
Array [
  Point {
    "_x": 0,
    "_y": 0,
  },
]
`;

exports[`RelativeMove new RelativeMove() 1`] = `
RelativeMove {
  "points": Array [
    Point {
      "_x": 0,
      "_y": 0,
    },
  ],
  "relative": false,
  "type": "m",
}
`;

exports[`RelativeMove point 1`] = `
Point {
  "_x": 0,
  "_y": 0,
}
`;

exports[`RelativeMove points 1`] = `
Array [
  Point {
    "_x": 0,
    "_y": 0,
  },
]
`;

exports[`RelativeMove values 1`] = `
Array [
  Point {
    "_x": 0,
    "_y": 0,
  },
]
`;
